### PR TITLE
base64 optimize decode

### PR DESCRIPTION
**Pull request by kostya**
after this commit crystal should win here: https://github.com/kostya/benchmarks#base64

### CODE REVIEW

File: src/base64.cr [hunk 1]
```diff
@@ -110,36 +113,49 @@ module Base64
...
-      yield (buf >> 8).to_u8
+      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
+      yield (a << 2 | b >> 4).to_u8
+      yield (b << 4 | c >> 2).to_u8
     elsif mod != 0
       raise Error.new("Wrong length")
     end
   end
 
+  private macro next_decoded_value
```
Comment (kostya) [inline src/base64.cr]: when i trying implement `next_decoded_value` as lambda, it much slower. why?

**Comment by asterite** [comment]
@kostya You are really amazing!! I just tried it on my machine and it's more than twice faster than before.

File: src/base64.cr (hunk 1 above)
Comment (asterite) [inline src/base64.cr]: If a lambda captures local variables, even if they don't escape the method, it becomes a closure and those variables are allocated in the heap. We might improve this in the future, but I think using a macro here is fine too as it guarantees that the performance will be optimal.

Comment (kostya) [inline src/base64.cr]: why not just inline closure without copying, if closure used only in this method.

Comment (asterite) [inline src/base64.cr]: For that we would need some sort of escape analysis. Or maybe for now we could track the variable that was assigned a closure and see if it's only invoked and never reassigned or passed to a method. But these are optimizations that I don't think we'll do soon.

Another alternative would be to allow defining nested defs, but that's something new too.

I think macros are perfectly fine for what you did. For example one thing that you can't do with normal methods is to return early from the parent caller. For example:

```crystal
def foo
  check_some_condition
end

def check_some_condition
  if some_condition
    # But I want to return from foo here!
    return
  end
end
```

With macros this is possible:

```crystal
def foo
  check_some_condition
end

macro check_some_condition
  if some_condition
    # But I want to return from foo here!
    return
  end
end
```

Also, in these cases the macro is just local to that class, so it's just a helper. Accessing outside variables like that isn't a big deal. Also, if you accidentally use incorrect names you will get a compile error, so it's safe.

**Pull request merged**